### ISSUE TITLE
isito-validation initContainer is not working in istio-cni-node

### ISSUE THREAD

**Issue by darkstarmv** [istio/istio#23457]
**Bug description**
- I have installed istio with Helm and everything works except of automatic sidecar injection
- I have tried Istio versions between 1.4.8, 1.5.0, 1.5.1 and 1.5.2, but same result
- with 1.4.8 version, pod get stuck in `Init:0/1` state
- with 1.5.X versions, pod get killed immediately
- going back to Istio 1.4.4 makes sidecar injection work and updating to 1.4.5 breaks it. so, there is some code that was introduced in 1.4.5 that broke it.
Looking at issues I see people describing same behavior in https://github.com/istio/istio/issues/21266

###  Setup details 
Kuberetes: `v1.15.11`
Istio CNI configuration on the server::
```
ls -alt /etc/cni/net.d/
total 8
drwxr-xr-x. 2 root root   68 May  1 13:08 .
-rw-r--r--. 1 root root  426 May  1 13:08 YYY-istio-cni.conflist
-rw-------. 1 root root 2824 May  1 13:08 ZZZ-istio-cni-kubeconfig
drwxr-xr-x. 3 root root   19 Apr 30 16:39 ..
```
Contents of the files:
```
cat  /etc/cni/net.d/*
{
  "plugins": [
    {
      "cniVersion": "0.3.1",
      "name": "istio-cni",
      "type": "istio-cni",
      "log_level": "debug",
      "kubernetes": {
        "kubeconfig": "/etc/cni/net.d/ZZZ-istio-cni-kubeconfig",
        "cni_bin_dir": "/opt/cni/bin",
        "exclude_namespaces": [
          "istio-system",
          "kube-system"
        ]
      }
    }
  ],
  "name": "k8s-pod-network",
  "cniVersion": "0.3.0"
}
# Kubeconfig file for Istio CNI plugin.
apiVersion: v1
kind: Config
clusters:
- name: local
  cluster:
    server: https://[192.168.128.1]:443
    certificate-authority-data: DDDDDDD
users:
- name: istio-cni
  user:
    token: "ZZZZZ"
contexts:
- name: istio-cni-context
  context:
    cluster: local
    user: istio-cni
current-context: istio-cni-context
```

When running with `1.4.8` I was able to capture Docker log for the pod 
`     "istio-iptables-go -…"   2 minutes ago       Up 2 minutes                            k8s_istio-validation_....`
Log:
```
in new validator: 172.30.22.223
Listening on 127.0.0.1:15001
Listening on 127.0.0.1:15006
Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
```
- CNI node pod log:
```
kubectl logs istio-cni-node-zzk9m repair-cni
{"level":"info","time":"2020-05-01T13:08:36.637040Z","msg":"Controller Option: Running as a Daemon."}
{"level":"info","time":"2020-05-01T13:08:36.637071Z","msg":"Controller Option: Labeling broken pods with label cni.istio.io/uninitialized=true"}
{"level":"info","time":"2020-05-01T13:08:36.637075Z","msg":"Filter option: Only managing pods with an annotation with key sidecar.istio.io/status"}
{"level":"info","time":"2020-05-01T13:08:36.637080Z","msg":"Filter option: Only managing pods with field selector spec.nodeName=SSSSSS,"}
{"level":"info","time":"2020-05-01T13:08:36.637084Z","msg":"Filter option: Only managing pods where init container is named istio-validation"}
{"level":"info","time":"2020-05-01T13:08:36.637088Z","msg":"Filter option: Only managing pods where init container exit status is 126"}
{"level":"info","time":"2020-05-01T13:13:50.917311Z","msg":"Pod detected as broken, adding label: sampleapp/sample-app-sox-sample-ap-deployment-5b56f4dff-zktlw"}
{"level":"info","time":"2020-05-01T13:13:50.917627Z","msg":"Labeling pod sampleapp/sample-app-sox-sample-ap-deployment-5b56f4dff-zktlw with label cni.istio.io/uninitialized=true"}
{"level":"info","time":"2020-05-01T13:13:51.048565Z","msg":"Pod detected as broken, adding label: sampleapp/sample-app-sox-sample-ap-deployment-5b56f4dff-zktlw"}
{"level":"info","time":"2020-05-01T13:13:51.048804Z","msg":"Pod sampleapp/sample-app-sox-sample-ap-deployment-5b56f4dff-zktlw already has label with key cni.istio.io/uninitialized, skipping"}
```

   

**Expected behavior**
- sidecar should be injected

**Steps to reproduce the bug**
- install istio
- configure auto injection on namespace
- deploy application

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
`istioctl version --remote
client version: 1.7-alpha.6d103c03fa30c66945843025aa029c9fd8517b67
control plane version: 1.4.8`

`kubectl version
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-19T11:13:54Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.11", GitCommit:"d94a81c724ea8e1ccc9002d89b7fe81d58f89ede", GitTreeState:"clean", BuildDate:"2020-03-12T21:00:06Z", GoVersion:"go1.12.17", Compiler:"gc", Platform:"linux/amd64"}`

`helm2 version
Client: &version.Version{SemVer:"v2.12.3", GitCommit:"eecf22f77df5f65c823aacd2dbd30ae6c65f186e", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.12.3", GitCommit:"eecf22f77df5f65c823aacd2dbd30ae6c65f186e", GitTreeState:"clean"}`

**How was Istio installed?**
- Istio installed with helm

**Environment where bug was observed (cloud vendor, OS, etc)**
- BYO K8s cluster baremetal and GCE

**Comment by darkstarmv** [comment]
Ran some more tests 1.4.4 works and going to 1.4.5 breaks it

**Comment by tariq1890** [comment]
I am facing this issue as well. What is the solution for this?

**Comment by stewartbutler** [comment]
So by the looks of these logs:
```
{"level":"info","time":"2020-05-01T13:08:36.637040Z","msg":"Controller Option: Running as a Daemon."}
{"level":"info","time":"2020-05-01T13:08:36.637071Z","msg":"Controller Option: Labeling broken pods with label cni.istio.io/uninitialized=true"}
```
it looks like you have the repair controller set to only label pods which are broken due to the Istio CNI race condition (https://github.com/istio/istio/issues/14327). Can you modify the env variable in the repair container of the istio-cni-node daemonset to turn on pod deletion and try again?

**Comment by howardjohn** [comment]
Any info @darkstarmv ?

**Comment by BrandonSharp** [comment]
I'm also having this issue. I've reverted back to not using CNI for the time being in my dev cluster, but would really like to get this enabled for other clusters with more stringent RBAC requirements if possible.

This is in an Azure AKS v1.16.10 cluster with an Operator deployment. Operator is in `istio-operator` namespace. IstioOperator deployment is as follows:

```yaml
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: istio-system
  name: istiocontrolplane
spec:
  profile: default
  tag: 1.6.5-distroless
  components:
    cni:
      enabled: true
      tag: 1.6.5
      namespace: kube-system
  addonComponents:
    grafana:
      enabled: true
  values:
    cni:
      excludeNamespaces:
      - istio-system
      - istio-operator
      - kube-system
      logLevel: info
```

It's perhaps worth noting that I had issues at first because the istio-cni-node pods were getting created in `istio-system`, and then another issue that there doesn't appear to be a distroless image for `install-cni`. After clearing up those with the above Operator spec and confirming the `istio-cni-node` pods were running on every node, I still ran into the issues described above.

```bash
❯ k logs -n my-ns my-pod istio-validation
in new validator: 10.244.2.195
Listening on 127.0.0.1:15001
Listening on 127.0.0.1:15006
Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
```

Logs for the repair container of the CNI pods were all similar:

```bash
{"level":"info","time":"2020-07-31T21:02:55.250121Z","msg":"Controller Option: Running as a Daemon."}
{"level":"info","time":"2020-07-31T21:02:55.250148Z","msg":"Controller Option: Deleting broken pods. Pod Labeling deactivated."}
{"level":"info","time":"2020-07-31T21:02:55.250155Z","msg":"Filter option: Only managing pods with an annotation with key sidecar.istio.io/status"}
{"level":"info","time":"2020-07-31T21:02:55.250159Z","msg":"Filter option: Only managing pods with field selector spec.nodeName=aks-agentpool-23039119-vmss000002,"}
{"level":"info","time":"2020-07-31T21:02:55.250163Z","msg":"Filter option: Only managing pods where init container is named istio-validation"}
{"level":"info","time":"2020-07-31T21:02:55.250168Z","msg":"Filter option: Only managing pods where init container exit status is 126"}
```


I'm happy to re-enable CNI and do some more digging to help get to the bottom of the issue, if needed, but it might be next week before I'm able to do so. Thanks!

**Comment by harpratap** [comment]
Facing the same issue for a while now, haven't caused any service disruption so not so critical for us.
It's difficult to reproduce, happens only on "large clusters" with a lot of activity ~1500 pods ~1000 virtualservices. Does not seem to have anything with resources though, it stays at ~30M memory and we have 100M requests for the repair-cni container. (It's the repair-cni that is getting killed)
`{"level":"info","time":"2020-08-03T05:27:36.470620Z","msg":"Filter option: Only managing pods where init container exit status is 126"}`

**Comment by stewartbutler** [comment]
If the init container is outputting ```Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused
Error connecting to 127.0.0.6:15002: dial tcp 127.0.0.1:0->127.0.0.6:15002: connect: connection refused``` it should eventually exit with an exit code 126. Are you not seeing that behavior? @harpratap @BrandonSharp

**Comment by sdake** [comment]
I have experienced part of this problem.

It is definitely clear that, even in 1.7 and master, the "validator" attempts to connect to 15002. Unfortunately, I am unable to find CNI design documentation and the code could use more comments. As a result, I am not sure what the validator does.

@stewartbutler if you need someone to run tests for you or provide more information, hit me up on slack or use a mention.

Cheers,
-steve

**Comment by howardjohn** [comment]
the validator will connect to port 15002 (could have been anything) and expect to be redirect to 15001 by our iptables. If this redirection doesn't happen it will fail.

**Comment by vikaschoudhary16** [comment]
I debugged this issue recently on GCE and EKS. This has nothing to do with port 15002 as @howardjohn  also pointed out. If this is happening means istio-cni could not setup redirection iptable rules. There are couple of scenarios depending on kubelet configuration which can result into failure of istio-cni loading and then all such cases will manifest in this istio-validation issue. Here are the ones that i could figure out:
1. on eks, kubelet is not running with `--cni-bin-dir`. This means kubelet will expect cni plugins at /opt/cni/bin(default). Istio manifests have default path as `/home/kubernetes/bin`. In kubelet logs, we can see `failed to find plugin "istio-cni" in path [/opt/cni/bin]`. This can be resolved by configuring `/opt/cni/bin/` as the cni binary path in the `IstioOperator` 
```
[root@ip-192-168-56-118 ~]# kubelet --version
Kubernetes v1.16.13-eks-ec92d4
[root@ip-192-168-56-118 ~]# ps ax | grep kubelet
 3844 ?        Ssl  278:24 /usr/bin/kubelet --cloud-provider aws --config /etc/kubernetes/kubelet/kubelet-config.json --kubeconfig /var/lib/kubelet/kubeconfig --container-
runtime docker --network-plugin cni --node-ip=192.168.56.118 --pod-infra-container-image=602401143452.dkr.ecr.eu-west-3.amazonaws.com/eks/pause:3.1-eksbuild.1 --node-label
s=eks.amazonaws.com/sourceLaunchTemplateVersion=1,alpha.eksctl.io/cluster-name=tsb-aws-ga,alpha.eksctl.io/nodegroup-name=default-pool,eks.amazonaws.com/nodegroup-image=ami
-0d1badbc7843bf2eb,eks.amazonaws.com/capacityType=ON_DEMAND,eks.amazonaws.com/nodegroup=default-pool,eks.amazonaws.com/sourceLaunchTemplateId=lt-090a540ce871ff5ed
 6479 pts/2    S+     0:00 grep --color=auto kubelet
[root@ip-192-168-56-118 ~]# 
```
2. On GKE(This was a 1.15 k8s cluster and istio was 1.6.12), kubelet runs with `--cni-bin-dir=/home/kubernetes/bin` which matches with the istio's default value of cniBinaryPath and thus path matches. But still this issue was there. I increased kubelet log verbosity from `--v=2` to `--v=6` and then i could see following in the logs:
```
1776:Nov 20 05:30:14 gke-firstcluster-default-pool-027e783f-482g kubelet[2721020]: I1120 05:30:14.664263 2721020 kubenet_linux.go:694] Adding default/nginx-7b66bb8c77-h8rfc to 'kubenet' with CNI 'bridge' plugin and runtime: &{ContainerID:5d56f57
```
 Above kubelet log means, kubelet is not loading `istio-cni` plugin rather loading the default `bridge` plugin. I could see cni config was there, specifying istio-cni, at the default cni config path i.e `/etc/cni/net.d` .  kubelet was running with `--network-plugin=kubenet`. In EKS also(1), `network-plugin` was `kubenet` and it was able to read the cni config. Though kubelet version was `1.16.3` on EKS. (may be in 1.15, kubenet does not load custom cni plugins, idk). When i switched kubelet config `--network-plugin=kubenet` to `--network-plugin=cni`, kubelet started loading `istio-cni`. But at pod bootstrap phase, kubelet logs were showing another error now:
```
21706:Nov 20 06:35:09 gke-firstcluster-default-pool-027e783f-482g kubelet[2780221]: W1120 06:35:09.103662 2780221 docker_sandbox.go:394] failed to read pod IP from plugin/docker: networkPlugin cni failed on the status hook for pod "details-v1-5b
c5dccd95-5d2n7_bookinfo": unexpected command output Device "eth0" does not exist.
```
Further, i could not debug because of time constraints and other priorities.

To summarize, this has to do with kubelet configs for cni and for the hints look at kubelet logs with --v=6

**Comment by darkstarmv** [comment]
I finally got some time to do more testing with `istio-cni` v1.8.1 and Kubernetes `v1.18.12`, but with the same result:
```
kubectl get po | grep cni
istio-cni-node-dzmqd                                              1/2     Running   0          24h
istio-cni-node-g22dt                                              1/2     Running   0          24h
istio-cni-node-qnxtl                                              1/2     Running   0          4h55m
istio-cni-node-xkz7t                                              1/2     Running   0          24h
istio-cni-node-zglr2                                              1/2     Running   0          4h55m
```
```
kubectl logs istio-cni-node-g22dt install-cni
2020-12-15T22:45:23.419013Z	info	Copied istio-cni to /host/opt/cni/bin.
2020-12-15T22:45:23.455129Z	info	Copied istio-cni-repair to /host/opt/cni/bin.
2020-12-15T22:45:23.493789Z	info	Copied istio-cni-taint to /host/opt/cni/bin.
2020-12-15T22:45:23.502839Z	info	Copied istio-iptables to /host/opt/cni/bin.
2020-12-15T22:45:23.502879Z	info	Directory /host/secondary-bin-dir is not writable, skipping.
2020-12-15T22:45:23.503808Z	info	Using CNI config template from CNI_NETWORK_CONFIG environment variable.
2020-12-15T22:45:23.503828Z	info	CNI config: {
  "cniVersion": "0.3.1",
  "name": "istio-cni",
  "type": "istio-cni",
  "log_level": "debug",
  "kubernetes": {
      "kubeconfig": "/etc/cni/net.d/ZZZ-istio-cni-kubeconfig",
      "cni_bin_dir": "/opt/cni/bin",
      "exclude_namespaces": [ "istio-system" ]
  }
}
```

```
kubectl logs istio-cni-node-g22dt repair-cni
{"level":"info","time":"2020-12-15T22:45:23.586611Z","msg":"Controller Option: Running as a Daemon."}
{"level":"info","time":"2020-12-15T22:45:23.586639Z","msg":"Controller Option: Deleting broken pods. Pod Labeling deactivated."}
{"level":"info","time":"2020-12-15T22:45:23.586643Z","msg":"Filter option: Only managing pods with an annotation with key sidecar.istio.io/status"}
{"level":"info","time":"2020-12-15T22:45:23.586647Z","msg":"Filter option: Only managing pods with field selector spec.nodeName=node1,"}
{"level":"info","time":"2020-12-15T22:45:23.586650Z","msg":"Filter option: Only managing pods where init container is named istio-validation"}
{"level":"info","time":"2020-12-15T22:45:23.586653Z","msg":"Filter option: Only managing pods where init container exit status is 126"}
```

```
kubectl describe po  istio-cni-node-g22dt
Name:                 istio-cni-node-g22dt
Namespace:            kube-system
Priority:             2000000000
Priority Class Name:  system-cluster-critical
Node:                 node1/x.x.x.x
Start Time:           Tue, 15 Dec 2020 16:45:22 -0600
Labels:               controller-revision-hash=698988f877
                      k8s-app=istio-cni-node
                      pod-template-generation=1
Annotations:          kubernetes.io/psp: 0-privileged
                      scheduler.alpha.kubernetes.io/critical-pod:
                      sidecar.istio.io/inject: false
Status:               Running
IP:                   x.x.x.x
IPs:
  IP:           x.x.x.x
Controlled By:  DaemonSet/istio-cni-node
Containers:
  install-cni:
    Container ID:  docker://4fe201b6814649bdabaca2edc3c324237473fd27d3a0312d0d1b1318dcf60507
    Image:         docker.io/istio/install-cni:1.8.1
    Image ID:      docker-pullable://istio/install-cni@sha256:97a4e6ce62181e8999bf7c9411cc5ef311ae26d59c983aba1505f53b7e73e7a1
    Port:          <none>
    Host Port:     <none>
    Command:
      install-cni
    State:          Running
      Started:      Tue, 15 Dec 2020 16:45:23 -0600
    Ready:          False
    Restart Count:  0
    Liveness:       http-get http://:8000/healthz delay=5s timeout=1s period=10s #success=1 #failure=3
    Readiness:      http-get http://:8000/readyz delay=0s timeout=1s period=10s #success=1 #failure=3
    Environment:
      CNI_NETWORK_CONFIG:  <set to the key 'cni_network_config' of config map 'istio-cni-config'>  Optional: false
      CNI_NET_DIR:         /etc/cni/net.d
      CHAINED_CNI_PLUGIN:  true
    Mounts:
      /host/etc/cni/net.d from cni-net-dir (rw)
      /host/opt/cni/bin from cni-bin-dir (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from istio-cni-token-54zn9 (ro)
  repair-cni:
    Container ID:  docker://acec1b3658bcd2ded7ac98a06adcb9e59493288f315b7faedfca6878325492fb
    Image:         docker.io/istio/install-cni:1.8.1
    Image ID:      docker-pullable://istio/install-cni@sha256:97a4e6ce62181e8999bf7c9411cc5ef311ae26d59c983aba1505f53b7e73e7a1
    Port:          <none>
    Host Port:     <none>
    Command:
      /opt/cni/bin/istio-cni-repair
    State:          Running
      Started:      Tue, 15 Dec 2020 16:45:23 -0600
    Ready:          True
    Restart Count:  0
    Environment:
      REPAIR_NODE-NAME:                (v1:spec.nodeName)
      REPAIR_LABEL-PODS:              true
      REPAIR_DELETE-PODS:             true
      REPAIR_RUN-AS-DAEMON:           true
      REPAIR_SIDECAR-ANNOTATION:      sidecar.istio.io/status
      REPAIR_INIT-CONTAINER-NAME:     istio-validation
      REPAIR_BROKEN-POD-LABEL-KEY:    cni.istio.io/uninitialized
      REPAIR_BROKEN-POD-LABEL-VALUE:  true
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from istio-cni-token-54zn9 (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  cni-bin-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /opt/cni/bin
    HostPathType:
  cni-net-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /etc/cni/net.d
    HostPathType:
  istio-cni-token-54zn9:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  istio-cni-token-54zn9
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  kubernetes.io/os=linux
Tolerations:     :NoSchedule
                 :NoExecute
                 CriticalAddonsOnly
                 node.kubernetes.io/disk-pressure:NoSchedule
                 node.kubernetes.io/memory-pressure:NoSchedule
                 node.kubernetes.io/network-unavailable:NoSchedule
                 node.kubernetes.io/not-ready:NoExecute
                 node.kubernetes.io/pid-pressure:NoSchedule
                 node.kubernetes.io/unreachable:NoExecute
                 node.kubernetes.io/unschedulable:NoSchedule
Events:
  Type     Reason     Age                     From                                    Message
  ----     ------     ----                    ----                                    -------
  Warning  Unhealthy  4m40s (x8731 over 24h)  kubelet, node1  Readiness probe failed: HTTP probe failed with statuscode: 503
```

**Comment by houshym** [comment]
@darkstarmv  I have same issue with following env:
1- rancher K3s  v1.20.6+k3s1
2- istio 1.8.3
following is istio-cni container logs

2021-04-21T06:59:43.621658Z	info	Copied istio-cni to /host/opt/cni/bin.

2021-04-21T06:59:43.659996Z	info	Copied istio-cni-repair to /host/opt/cni/bin.

2021-04-21T06:59:43.697399Z	info	Copied istio-cni-taint to /host/opt/cni/bin.

2021-04-21T06:59:43.704551Z	info	Copied istio-iptables to /host/opt/cni/bin.

2021-04-21T06:59:43.704585Z	info	Directory /host/secondary-bin-dir is not writable, skipping.

2021-04-21T06:59:43.705227Z	info	Using CNI config template from CNI_NETWORK_CONFIG environment variable.

2021-04-21T06:59:43.705250Z	info	CNI config: {

  "cniVersion": "0.3.1",

  "name": "istio-cni",

  "type": "istio-cni",

  "log_level": "info",

  "kubernetes": {

      "kubeconfig": "/etc/cni/net.d/ZZZ-istio-cni-kubeconfig",

      "cni_bin_dir": "/opt/cni/bin",

      "exclude_namespaces": [ "istio-system", "kube-system" ]

  }

}

**Comment by houshym** [comment]
any update on this @howardjohn
